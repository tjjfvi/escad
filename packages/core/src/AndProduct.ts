import { Product, _Product, checkTypeProperty, ProductType } from "."

export interface AndProduct<T extends readonly Product[] = readonly Product[]> extends _Product {
  readonly type: "AndProduct",
  readonly children: T,
}

export const AndProduct = {
  create: <T extends readonly Product[]>(children: T): AndProduct<T> => ({
    type: "AndProduct",
    children,
  }),
  isAndProduct: checkTypeProperty.string<AndProduct>("AndProduct"),
  getAndProductType: <T extends AndProduct>(product: T): AndProductType<T> =>
    AndProductType.create(product.children.map(Product.getProductType)) as any,
}

export interface AndProductType<T extends AndProduct = AndProduct> {
  readonly type: "AndProductType",
  readonly elementTypes: (
    T extends AndProduct<infer U>
      ? { [K in keyof U]: U[K] extends Product ? ProductType<U[K]> : never }
      : never
  ),
}

export const AndProductType = {
  create: (elementTypes: ProductType[]): AndProductType => ({
    type: "AndProductType",
    elementTypes,
  }),
  isAndProductType: checkTypeProperty.string<AndProductType>("AndProductType"),
  getAndProductType: <T extends AndProduct>(product: T): AndProductType<T> =>
    AndProductType.create(product.children.map(x => Product.getProductType(x))) as any,
}
