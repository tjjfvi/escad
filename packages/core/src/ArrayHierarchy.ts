
import { checkTypeProperty } from "./checkTypeProperty";
import { _Hierarchy, Hierarchy } from "./Hierarchy";

export interface ArrayHierarchy extends _Hierarchy {
  readonly type: "ArrayHierarchy",
  readonly children: Hierarchy[],
}

export const ArrayHierarchy = {
  create: ({
    children,
    linkedProducts = children.flatMap(x => x.linkedProducts ?? []),
  }: Omit<ArrayHierarchy, "type">): ArrayHierarchy => ({
    type: "ArrayHierarchy",
    children,
    linkedProducts,
  }),
  from: (array: unknown[], raw = false) =>
    ArrayHierarchy.create({
      children: array.map(x => Hierarchy.from(x, raw))
    }),
  isArrayHierarchy: checkTypeProperty.string<ArrayHierarchy>("ArrayHierarchy"),
}
