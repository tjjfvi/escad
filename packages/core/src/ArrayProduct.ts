
import { checkTypeProperty } from "./checkTypeProperty";
import { Product, _Product, ProductType } from "./Product";

export interface ArrayProduct<T extends Product = Product> extends _Product {
  readonly type: "ArrayProduct",
  readonly children: readonly T[],
}

export const ArrayProduct = {
  create: <T extends Product>(children: readonly T[]): ArrayProduct<T> => {
    if(children.length === 0)
      throw new Error("ArrayProducts cannot be empty");
    return {
      type: "ArrayProduct",
      children,
    };
  },
  isArrayProduct: checkTypeProperty<ArrayProduct>("ArrayProduct"),
  getArrayProductType: <T extends ArrayProduct>(product: T): ArrayProductType<T> =>
    ArrayProductType.create(Product.getProductType(product.children[0])) as any
}

export interface ArrayProductType<T extends ArrayProduct = ArrayProduct> {
  readonly type: "ArrayProductType",
  readonly elementType: ProductType<T extends ArrayProduct<infer U> ? U : never>,
}

export const ArrayProductType = {
  create: (elementType: ProductType): ArrayProductType => ({
    type: "ArrayProductType",
    elementType,
  }),
  isArrayProductType: checkTypeProperty<ArrayProductType>("ArrayProductType"),
}
