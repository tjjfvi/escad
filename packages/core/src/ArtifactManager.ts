
import fs from "fs-extra";
import { Readable } from "stream";

import { Sha } from "./hash";
import { ReadonlyArtifactManager } from "./ReadonlyArtifactManager";

export abstract class ArtifactManager<T> extends ReadonlyArtifactManager<T> {

  abstract deserialize(stream: Readable): Promise<T | null>;

  async lookup(sha: Sha){
    return await this.cache.getAsync(sha.hex, async () => {
      const path = await this.getPath(sha);
      let stream = fs.createReadStream(path);
      return await new Promise(resolve => {
        stream
          .once("open", () => {
            resolve(this.deserialize(stream));
          })
          .once("error", () => resolve(null));
      })
    })
  }

  abstract getSha(t: T): Promise<Sha>

  reference = () => Sha.reference().map<T>({
    serialize: t => this.getSha(t),
    deserialize: async sha => {
      const artifact = await this.lookup(sha);
      if(!artifact)
        throw new Error(`Could not find artifact of sha ${sha}`);
      return artifact;
    }
  })

}
