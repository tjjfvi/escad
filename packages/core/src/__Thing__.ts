
import { ExtensibleFunction } from "./ExtensibleFunction";

export class __Thing__ extends ExtensibleFunction {

  declare protected __thing__: true;

}
