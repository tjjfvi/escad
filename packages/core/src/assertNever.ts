
import { inspect } from "util"

export function assertNever(
  value: never,
  message: (str: string) => string = str => `Expected never, got ${str}`,
): never{
  throw new Error(message(inspect(value)))
}
