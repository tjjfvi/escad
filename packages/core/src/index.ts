
export * from "./builtins";
import builtins from "./builtins";
import Element, { Elementish } from "./Element";
import Product from "./Product";

const escadFunc = <T extends Product>(...a: Elementish<T>[]) => {
  return new Element(a.length === 1 ? a[0] : a);
}

const escad = new Proxy(escadFunc, {
  get: (target, key) =>
    key in target ?
      target[key as keyof typeof target] :
      builtins[key as keyof typeof builtins]
}) as typeof escadFunc & typeof builtins;

export default escad;

export { default as Component } from './Component';
export { default as Element } from './Element';
export { default as Hierarchy } from './Hierarchy';
export { default as Id } from './Id';
export { default as Operation } from './Operation';
export { default as Product } from './Product';
export { default as Work } from './Work';
export { default as ProductManager } from './ProductManager';
export * from './Component';
export * from './Element';
export * from './Hierarchy';
export * from './Id';
export * from './Operation';
export * from './Product';
export * from './Work';
export * from "./builtins";
export * from "./mapOperation";
export * from "./ExportType";

