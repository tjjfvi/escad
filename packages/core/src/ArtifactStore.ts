
import { ArtifactManager } from "./ArtifactManager"
import { Hash } from "./Hash"
import { WrappedValue } from "./WrappedValue"

export interface ArtifactStore {
  readonly storeRaw?: (hash: Hash<unknown>, artifact: WrappedValue, artifactManager: ArtifactManager) => Promise<void>,
  readonly storeRef?: (loc: readonly unknown[], hash: Hash<unknown>, artifactManager: ArtifactManager) => Promise<void>,
  readonly lookupRaw?: (hash: Hash<unknown>, artifactManager: ArtifactManager) => Promise<WrappedValue | null>,
  readonly lookupRef?: (loc: readonly unknown[], artifactManager: ArtifactManager) => Promise<WrappedValue | null>,
}
