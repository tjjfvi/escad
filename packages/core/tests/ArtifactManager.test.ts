
import { ArtifactManager, ArtifactStore, Hash, Id } from "../src";

const createArtifactStoreMock = (name: string, output: unknown[], returnValues: unknown[]): ArtifactStore => ({
  lookupRaw: async hash => {
    output.push([name, "lookupRaw", hash])
    return returnValues.pop() ?? null
  },
  lookupRef: async loc => {
    output.push([name, "lookupRef", loc])
    return returnValues.pop() ?? null
  },
  storeRaw: async (hash, buffer) => {
    output.push([name, "storeRaw", hash, buffer.toString("base64")])
  },
  storeRef: async (loc, hash) => {
    output.push([name, "storeRef", loc, hash])
  },
})

test("", async () => {
  const artifactManager = new ArtifactManager();
  const output = [] as unknown[]
  const store0 = createArtifactStoreMock("store0", output, []);
  const store1 = createArtifactStoreMock("store1", output, [
    Buffer.from(JSON.stringify({ test0: true })),
    Buffer.from(Hash.create({ test1: true }), "hex"),
    { test2: true },
    { test3: true },
  ]);
  const excludeStores = new Set([store1]);
  artifactManager.artifactStores.push({}, store0, store1);
  const artifacts = [
    { isTestArtifact: true },
    Buffer.from(Hash.create("test"), "hex"),
    Id.create(__filename, "@escad/core", "Test", "ArtifactTest", "0"),
  ];
  for(const artifact of artifacts) {
    await artifactManager.lookupRaw(Hash.create(artifact));
    await artifactManager.lookupRaw(Hash.create(artifact), excludeStores);
    await artifactManager.storeRaw(artifact);
    await artifactManager.storeRaw(Promise.resolve(artifact), excludeStores);
  }
  await artifactManager.lookupRef(artifacts);
  await artifactManager.lookupRef(artifacts, excludeStores);
  await artifactManager.storeRef(artifacts, artifacts[0]);
  await artifactManager.storeRef([], artifacts[1], excludeStores);
  expect(output).toMatchSnapshot();
})
