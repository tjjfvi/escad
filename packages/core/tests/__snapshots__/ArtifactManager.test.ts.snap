// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Array [
    "store0",
    "lookupRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
  ],
  Array [
    "store1",
    "lookupRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
  ],
  Array [
    "store0",
    "lookupRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
  ],
  Array [
    "store0",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    Object {
      "value": Object {
        "isTestArtifact": true,
      },
    },
  ],
  Array [
    "store1",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    Object {
      "value": Object {
        "isTestArtifact": true,
      },
    },
  ],
  Array [
    "store0",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    Object {
      "value": Object {
        "isTestArtifact": true,
      },
    },
  ],
  Array [
    "store0",
    "lookupRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
  ],
  Array [
    "store1",
    "lookupRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
  ],
  Array [
    "store0",
    "lookupRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
  ],
  Array [
    "store0",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    Object {
      "value": "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
  Array [
    "store1",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    Object {
      "value": "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
  Array [
    "store0",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    Object {
      "value": "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
  Array [
    "store0",
    "lookupRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
  ],
  Array [
    "store1",
    "lookupRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
  ],
  Array [
    "store0",
    "lookupRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
  ],
  Array [
    "store0",
    "storeRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
    Object {
      "value": "@escad/core/Test/ArtifactTest",
    },
  ],
  Array [
    "store1",
    "storeRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
    Object {
      "value": "@escad/core/Test/ArtifactTest",
    },
  ],
  Array [
    "store0",
    "storeRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
    Object {
      "value": "@escad/core/Test/ArtifactTest",
    },
  ],
  Array [
    "store0",
    "lookupRef",
    Array [
      Object {
        "isTestArtifact": true,
      },
      "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
      "@escad/core/Test/ArtifactTest",
    ],
  ],
  Array [
    "store1",
    "lookupRef",
    Array [
      Object {
        "isTestArtifact": true,
      },
      "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
      "@escad/core/Test/ArtifactTest",
    ],
  ],
  Array [
    "store0",
    "storeRaw",
    "6d861320c1c2dd822c9d766782f0f6144cf66ab818f601ba4466493bbdf76778",
    Object {
      "value": Object {
        "test0": true,
      },
    },
  ],
  Array [
    "store1",
    "storeRaw",
    "6d861320c1c2dd822c9d766782f0f6144cf66ab818f601ba4466493bbdf76778",
    Object {
      "value": Object {
        "test0": true,
      },
    },
  ],
  Array [
    "store0",
    "lookupRef",
    Array [
      Object {
        "isTestArtifact": true,
      },
      "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
      "@escad/core/Test/ArtifactTest",
    ],
  ],
  Array [
    "store0",
    "storeRef",
    Array [
      Object {
        "isTestArtifact": true,
      },
      "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
      "@escad/core/Test/ArtifactTest",
    ],
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
  ],
  Array [
    "store1",
    "storeRef",
    Array [
      Object {
        "isTestArtifact": true,
      },
      "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
      "@escad/core/Test/ArtifactTest",
    ],
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
  ],
  Array [
    "store0",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    Object {
      "value": Object {
        "isTestArtifact": true,
      },
    },
  ],
  Array [
    "store1",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    Object {
      "value": Object {
        "isTestArtifact": true,
      },
    },
  ],
  Array [
    "store0",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    Object {
      "value": Object {
        "isTestArtifact": true,
      },
    },
  ],
  Array [
    "store1",
    "storeRaw",
    "9057425f657d25f62d1adaf518df899b1703b0694e6f3e0a2ea64457be49b90f",
    Object {
      "value": Object {
        "isTestArtifact": true,
      },
    },
  ],
  Array [
    "store0",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    Object {
      "value": "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
  Array [
    "store1",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    Object {
      "value": "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
  Array [
    "store0",
    "storeRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
    Object {
      "value": "@escad/core/Test/ArtifactTest",
    },
  ],
  Array [
    "store1",
    "storeRaw",
    "fbbc3210f76789ad2ef86aec42dc8b3a296acd023029768e82ae7704e2ec28a6",
    Object {
      "value": "@escad/core/Test/ArtifactTest",
    },
  ],
  Array [
    "store0",
    "storeRef",
    Array [],
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
  ],
  Array [
    "store0",
    "storeRaw",
    "6e9c3da9ea693b2a19048fc7312e0c877018aa0ee5195867051372f2e58dadd0",
    Object {
      "value": "0a843fe2c3373dd9f5072dc37674ccab20eb807c564ca34b6384e4cae0f318c2",
    },
  ],
]
`;
